### PR TITLE
Home work4

### PR DESCRIPTION
**Pull request by Allkman**

### CODE REVIEW

File: Task1.GenericDiagonalMatrix/Entities/DiagonalMatrix.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                for (int column = 0; column < MatrixArray.Length; column++)
+                {
+                    sb.Append(this[row, column]);
+                }
+            }
+            return sb.ToString();
+        }
+        protected virtual void OnElementChanged(ElementChangedEventArgs<T> e)
+        {
+            for (int i = 0; i < MatrixArray.Length; i++)
```
Comment (radislavkireev) [inline Task1.GenericDiagonalMatrix/Entities/DiagonalMatrix.cs]: what is the purpose of that "for"?
Drop it and  raise event when element changed

File: Task1.GenericDiagonalMatrix/Entities/DiagonalMatrix.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                }
+                else if (i == j)
+                {
+                    return MatrixArray[i];
+                }
+                else return default(T);
+            }
+            set
+            {
+                if (i >= 0 && j >= 0 || i <= _size && j <= _size || i == j)
```
Comment (radislavkireev) [inline Task1.GenericDiagonalMatrix/Entities/DiagonalMatrix.cs:37]: for setter IndexOutOfRangeException should be thrown under the same conditions as for getter

File: Task1.GenericDiagonalMatrix/Helpers/GenericMatrixExtension.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿namespace Task1.GenericDiagonalMatrix
+{
+    internal static class GenericMatrixExtension
+    {
+        public static DiagonalMatrix<T> Add<T>(
+            this DiagonalMatrix<T> first,
+            DiagonalMatrix<T> second,
+            Func<T?, T?, T> AdditionOfT)
```
Comment (radislavkireev) [inline Task1.GenericDiagonalMatrix/Helpers/GenericMatrixExtension.cs]: camel case - **a**dditionOfT

File: Task1.GenericDiagonalMatrix/Helpers/GenericMatrixExtension.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+                newArrayLength = first.Size;
+            }
+            else if (second.Size > first.Size)
+            {
+                newArrayLength = second.Size;
+            }
+            var tempMatrix = new DiagonalMatrix<T>(newArrayLength);
+            for (int i = 0; i < newArrayLength; i++)
+            {
+                tempMatrix[i,i] = AdditionOfT.Invoke(first[i,i], second[i,i]);
```
Comment (radislavkireev) [inline Task1.GenericDiagonalMatrix/Helpers/GenericMatrixExtension.cs]: per requirements setter should throw IndexOutOfRange when i is out of range, so the code should handle this case.

**Review by radislavkireev — CHANGES_REQUESTED** [review]

File: Task1.GenericDiagonalMatrix/Entities/DiagonalMatrix.cs (hunk 2 above)
Comment (Allkman) [reply Task1.GenericDiagonalMatrix/Entities/DiagonalMatrix.cs:37]: In the set{  } ?
if (i < 0 && j < 0 || i >= _size && j >= _size)
   {
       throw new IndexOutOfRangeException();
   }

**Review by radislavkireev — APPROVED** [review]

**Pull request closed without merging**